### PR TITLE
Proposed recommendations for authoring tasks

### PR DESCRIPTION
**Pull request by mogsie**
# Changes

I added some recommendations after some months of working with Tekton Pipelines, as recommended by @ImJasonH in the Tekton slack.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

(removed N/A checkboxes)

### CODE REVIEW

File: recommendations.md
Comment (chmouel) [inline recommendations.md:87]: I think :

```bash
var=$(cat <<EOF
$(params.one)
EOF
)
```

should work too

File: recommendations.md
Comment (chmouel) [inline recommendations.md:164]: FYI: this is how we do it in Python : 

https://github.com/tektoncd/catalog/blob/master/task/github-add-comment/0.2/github-add-comment.yaml#L76-L78

It has the same issue as shell, if you use `var = "$(params.foo)"` and params.foo has a " then you run into problems

**Review by chmouel — APPROVED** [review]
Sounds good to me, maybe we want to add that other languages can have this issue too!

@Divyansh42 @vinamra28 we have that problem too in the catalog, for example this task https://github.com/tektoncd/catalog/blob/master/task/send-to-channel-slack/0.1/send-to-channel-slack.yaml#L32 is using shell and params, so if you have a " in there (which is probably in a message) it would fail. We may want to fix it as recommended here

File: recommendations.md
Comment (mogsie) [reply recommendations.md:164]: Yes, I'll 
* add a python example of the escaping issue
* add a pyton example on how to invoke a python script!

File: recommendations.md
Comment (mogsie) [reply recommendations.md:87]: Almost.  Shell escape rules are treacherous:
* If I send in `$(cat)` as the parameter, the script will evaluate `cat` which will likely fail because there is nothing in stdin
* If I send in "EOF" then the parameter will fail, and have the value EOF.
I'm pretty sure that there are no "good" interpolation rules for shell (or any language for that matter); because tekton just doesn't know (and rightly so) about the chosen language's escaping rules.

So in my opinion, the rule of thumb should be to avoid it unless you're very certain about the data being passed in.

**Comment by mogsie** [comment]
Sorry about the noise, I used the wrong e-mail address when committing, causing the EasyCLA to misfire.  I can probably expand on some sections, or remove some if you don't feel they should be recommendations.

* I moved the bit about using other languages _first_ so that I could use python examples further down.
* I expanded on the interpolation bit by adding a python example
* I also added a section about security issues, since that was my main concern with interpolation.

File: recommendations.md
Comment (chmouel) [reply recommendations.md]: that's right 👍🏽

File: recommendations.md
Comment (chmouel) [inline recommendations.md:79]: nit: maybe python3 :)

(/usr/bin/python is not found by default on fedora/ubuntu)

**Comment by chmouel** [comment]
/lgtm

Thanks for updating this document 👍🏽

**Comment by chmouel** [comment]
I didn't meant to +2A directly and not waiting for another reviewer review, sorry about this

File: recommendations.md
Comment (vinamra28) [inline recommendations.md:196]: hi @mogsie, I agree with this point that it becomes easy to write a script and being able to test that locally but what I was thinking was will this not make the task mutable and as per [TEP-0003](https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md) we should make the task immutable so that we don't break experience of the users using the tasks from the catalog.
Also the other question which I have in my mind is where to keep the script that ?

Comment (mogsie) [reply recommendations.md:196]: To be honest, I submitted these as general Task authoring guidelines; not guidelines for the Catalog per se.  Maybe these guidelines should be moved somewhere more appropriate (but it makes sense that the catalog espouses the recommendations, wherever they are).

But as for immutability, checking in a Task and ConfigMap to the catalog won't gain any of the benefits of extracting the script to its own file; not easily tested, edited, shellchecked, etc.  Committing it as a _kustomize_ manifest is of course an alternative.  Internally we use kustomize for these things, but I felt it was a stretch to have a recommendation to use kustomize.  It would of course be possible to provide a kustomize manifest in the catalog, and have simple instructions to create the ConfigMap by hand (if kustomize is not desirable) `--from-file` as described in the task.

**Pull request merged**
